### PR TITLE
#200 password reset

### PR DESCRIPTION
**Pull request by Mehrdadmaskull**
This pull request has the endpoint needed for the user to reset their password. "env" file needs to be updated with the DOMAIN name and also the API key.

### CODE REVIEW

File: Server/src/models/AuthenticationModel.ts [hunk 1]
```diff
@@ -45,6 +45,22 @@ export class AuthenticationModel {
         return userEmail !== undefined;
     }
 
+    /*
+     * Method to verify if the user with the token exists.
+     * @param email - User's email
+     * @param token - User's token from URL
+     */
+    static async verifyResetToken(email: string, token: string): Promise<boolean> {
```
Comment (GD-Dheer) [inline Server/src/models/AuthenticationModel.ts:53]: Cleanly created

File: Server/src/models/AuthenticationModel.ts [hunk 2]
```diff
@@ -116,10 +133,23 @@ export class AuthenticationModel {
...
+    static async resetPassword(email: string, token: string) {
 
         let connection = getConnection();
+
+        await connection.manager
+            .createQueryBuilder(Accounts, 'accounts')
+            .update('accounts')
+            .set({ resetToken: token })
+            .where('accounts.email = :email', { email: email })
+            .execute()
```
Comment (GD-Dheer) [inline Server/src/models/AuthenticationModel.ts:150]: Does this let the user set the new token? Where is the user inputted password? Is it contained inside resettoken?

File: Server/src/routes/authenticationRouter.ts [hunk 3]
```diff
@@ -19,9 +19,8 @@ router.post('/login', (request: Request, response: Response, next: NextFunction)
     authenticationController.createLoginRequest(request, response, next);
 });
 
-//TODO: Implement when doing password reset 
-router.get('/resetPassword', (request: Request, response: Response, next: NextFunction) => {
-
+router.post('/resetPassword', JWTAuthenticator.verifyJWT, async (request: Request, response: Response, next: NextFunction) => {
```
Comment (GD-Dheer) [inline Server/src/routes/authenticationRouter.ts]: Can you change this to /api/v1/resetpassword

File: Server/src/services/authenticationService.ts [hunk 4]
```diff
@@ -2,17 +2,20 @@ import 'dotenv/config';
...
 import { AuthenticationModel } from '../models/AuthenticationModel'
-import { IFetchUserDetail } from '../genericInterfaces/AuthenticationInterfaces';
+import { IFetchUserDetail, IPasswordResetInformation } from '../genericInterfaces/AuthenticationInterfaces';
 import { IJwtParams } from '../genericInterfaces/AuthenticationInterfaces';
 import { ILoginInformation } from '../genericInterfaces/AuthenticationInterfaces';
 import { IResponse } from '../genericInterfaces/ResponsesInterface'
 import { ISignUpInformation } from '../genericInterfaces/AuthenticationInterfaces';
 import { IUpdateUserDetail } from './../genericInterfaces/AuthenticationInterfaces';
 
+let mg = mailgun({ apikey: process.env.MAILGUN_API_KEY, domain: process.env.DOMAIN_NAME });
```
Comment (GD-Dheer) [inline Server/src/services/authenticationService.ts]: This is causing travis fails

File: Server/src/services/authenticationService.ts [hunk 5]
```diff
@@ -89,6 +92,53 @@ export class AuthenticationService {
...
+            if (!resetTokenChanged) {
+                throw new InternalServerError("Internal server error when executing set resetToken on user");
+            }
+
+            let data = {
+                from: "noreply@databoom.com",
+                to: resetPasswordInformation.email,
+                subject: 'Reset your password',
+                html: `
+                    <h2>To reset your password, please click on this link</h2>
```
Comment (GD-Dheer) [inline Server/src/services/authenticationService.ts:124]: A+

**Review by GD-Dheer — APPROVED** [review]

File: Server/src/models/AuthenticationModel.ts (hunk 2 above)
Comment (Mehrdadmaskull) [reply Server/src/models/AuthenticationModel.ts:150]: No this is the first step, to assign a user a temp resetToken.. another method will be added for when the user actually changes their password

File: Server/src/controllers/authenticationController.ts [hunk 6]
```diff
@@ -51,6 +51,23 @@ export class AuthenticationController {
...
+        else {
+            let requestParams: any = { ...request.body };
+            let passwordResetInfo: IPasswordResetInformation = requestParams;
+            let res: any = await this.callServiceForPasswordReset(passwordResetInfo, response, next);
+            return res;
+        }
+    }
+
+    private validatePasswordResetRequest(request: Request): boolean {
+        return !request.body.hasOwnProperty('email');
```
Comment (ds161833) [inline Server/src/controllers/authenticationController.ts]: shouldnt this be inverted? if email is present this will invalidate the request?

Comment (ds161833) [reply Server/src/controllers/authenticationController.ts]: also change this to return request.body.email

File: Server/src/controllers/authenticationController.ts [hunk 7]
```diff
@@ -123,6 +140,20 @@ export class AuthenticationController {
         }
     }
 
+    private async callServiceForPasswordReset(passwordResetInfo: IPasswordResetInformation, response: Response, next: NextFunction): Promise<Response> {
+        this.authenticationService = new AuthenticationService();
+        let serviceResponse: IResponse
+        try {
+            serviceResponse = await this.authenticationService.resetPassword(passwordResetInfo);
+
```
Comment (ds161833) [inline Server/src/controllers/authenticationController.ts]: blanks

File: Server/src/services/authenticationService.ts (hunk 5 above)
Comment (ds161833) [reply Server/src/services/authenticationService.ts:124]: nice

**Review by ds161833 — CHANGES_REQUESTED** [review]
checkout the email validation comment

File: Server/src/services/authenticationService.ts (hunk 4 above)
Comment (Mehrdadmaskull) [reply Server/src/services/authenticationService.ts]: It's supposed to be apiKey, fixed

File: Server/src/controllers/authenticationController.ts (hunk 6 above)
Comment (Mehrdadmaskull) [reply Server/src/controllers/authenticationController.ts]: Na, can't be inverted, the value returned is stored in `invalidResponse`, so should be good

**Review by ds161833 — APPROVED** [review]

**Pull request merged**